src/t_zset.c

```diff
@@ -3760,16 +3766,36 @@ void genericZpopCommand(client *c, robj **keyv, int keyc, int where, int emitkey
         return;
     }
 
-    void *arraylen_ptr = addReplyDeferredLen(c);
     long result_count = 0;
 
-    /* We emit the key only for the blocking variant. */
-    if (emitkey) addReplyBulk(c,key);
+    /* Respond with a single (flat) array in RESP2 or if count is 0
+     * (returning a single element). In RESP3, when count > 0 use nested array. */
+    int use_nested_array = (c->resp > 2 && count != 0);
+
+    /* When count is 0, we need to correct it to 1 for plain single pop. */
+    if (count == 0) count = 1;
 
-    /* Respond with a single (flat) array in RESP2 or if countarg is not
-     * provided (returning a single element). In RESP3, when countarg is
-     * provided, use nested array.  */
-    int use_nested_array = c->resp > 2 && countarg != NULL;
+    long llen = zsetLength(zobj);
+    long rangelen = (count > llen) ? llen : count;
+
+    /* need a comment... */
+    if (use_nested_array_with_key) {
+        /* ZMPOP/BZMPOP in RESP2 and RESP3. */
+        addReplyArrayLen(c, 2);
+    } else {
+        if (use_nested_array) {
+            /* ZPOPMIN/ZPOPMAX with COUNT option in RESP3. */
+            addReplyArrayLen(c, rangelen + (emitkey != 0));
+        } else {
+            /* BZPOPMIN/BZPOPMAX in RESP2 and RESP3 or ZPOPMIN/ZPOPMAX in RESP2. */
+            addReplyArrayLen(c, rangelen * 2 + (emitkey != 0));
+        }
+    }
+
+    /* We emit the key only for the blocking variant or in ZMPOP/BZMPOP. */
+    if (emitkey || use_nested_array_with_key) addReplyBulk(c,key);
+
+    if (use_nested_array_with_key) addReplyArrayLen(c, rangelen);
```

Comment:
now that you'll have just two variables (`use_nested_array` and `emitkey`), maybe this code will be easier?
in theory there are 4 variants then. one ()can be an assertion.

p.s. comments like you added, specifying an example of which command this case will be used, are good to make the code easier to read..

let me know if i'm missing something..

Reply:
Right, i think now the code is easier, thanks for the suggestion